psci reset with wdog

MrVan opened the pull request:
This is to support psci reset on i.MX. The wdog is shared between normal world and secure world. The secure world needs to parse the wdog node and check whether `fsl,ext-reset-output` exists or not, then decide to how to trigger wdog reset. imx_wdog_restart is exposed to imx psci reset.

Introduced get_dt_blob, dt_read_bool,CFG_MMAP_REGIONS.

CFG_MMAP_REGIONS maybe not a good way here, but the easiest way to fix my issue. core_mmu_add_mapping may fail, because RES_VASPACE entry in static memmap maybe used by others, such as device tree.

A few typo fixes.

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -726,6 +740,8 @@ static void init_fdt(unsigned long phys_fdt)
 		panic();
 	}
+
+	dt_blob_addr = fdt;
```
This pointer isn't valid forever. It should be cleared when OP-TEE has initialized and exits the first time.

MrVan replied:
I need to parse the dtb to get fsl,ext-reset-output in wdog node. In this patchset, I use driver_init(xxx_wdog_xx) to handle that. If not expose dtb here, I am not able to parse dtb for the drivers. Could we expose fdt in init_fdt, and destroy the pointer when first back to normal world? Or we create a copy of the dtb in optee secure side?

jenswi-linaro replied:
If you clear it before `init_primary_helper()` (after calling `init_teecore()`) it should be OK.

MrVan replied:
Thanks. I'll clear it as you suggested.

jenswi-linaro commented on core/drivers/imx_wdog.c:
```diff
@@ -0,0 +1,159 @@
+	uint32_t i;
+
+#ifdef CFG_MX7
+	const char *wdog_path[4] = {
```
Could have used `[]` instead of `[4]`

MrVan replied:
Fix in next version

MrVan commented on core/include/kernel/dt.h:
```diff
@@ -89,6 +89,16 @@ const struct dt_driver *__dt_driver_end(void);
+ *
+ * Returs true on success or false if no propname.
+ */
+bool dt_read_bool(const void *fdt, int offs, const char *propname);
```
Fix in next version

MrVan commented on core/kernel/dt.c:
```diff
@@ -58,6 +58,15 @@ const struct dt_driver *__dt_driver_end(void)
+
+	prop = fdt_getprop(fdt, offs, propname, NULL);
+
+	return prop ? true : false;
```
I would like to use true/false as return value, this is to check whether a property exist or not.

MrVan commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -684,6 +696,8 @@ static void init_fdt(unsigned long phys_fdt)
 	void *fdt;
 	int ret;
 
+	dt_blob_addr = NULL;
```
The dt_blob_addr is not a static variable. So init it with value NULL in init_fdt first.

MrVan replied:
ok. Will fix it.

MrVan commented on core/kernel/dt.c:
```diff
@@ -144,14 +144,17 @@ static paddr_t _fdt_read_paddr(const uint32_t *cell, int n)
 	const void *reg;
-	int ncells;
-	int len;
+	int parent, ncells, len;
```
ok

MrVan commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -51,7 +51,7 @@
 #include "core_mmu_private.h"
 
-#define MAX_MMAP_REGIONS	13
+#define MAX_MMAP_REGIONS	CFG_MMAP_REGIONS
```
ok, fix in next version

MrVan commented on core/drivers/imx_wdog.c:
```diff
@@ -0,0 +1,159 @@
+	write16(val, wdog_base + WCR_OFF);
+	write16(val, wdog_base + WCR_OFF);
+
+	asm volatile ("b .\r\n");
```
ok

MrVan commented:
Updated:
proerty->property
Returs->Returns
Spilt varaiable definition into different lines.
Allow platform set MAX_MMAP_REGIONS.
asm volatile("b .") -> while(1);
dt_read_bool->dt_have_prop
set dt_blob_addr with NULL after init_teecore
const char *wdog_path[4] -> static const char * const wdog_path[];

jenswi-linaro commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -80,6 +80,9 @@ endif
+CFG_IMX_WDOG ?= y
 endif
 
+MAX_MMAP_REGIONS ?= 24
```
This variable should have a `CFG_` prefix, also the config magic is only applied to `CFG_` variables if I'm not misstaken.

MrVan replied:
Per @etienne-lms 's comments:
```
I think the platform_config.h could set MAX_MMAP_REGIONS and change here:

+#ifndef MAX_MMAP_REGIONS
 #define MAX_MMAP_REGIONS	13
+#endif
Main point is that this config information does not really need to be CFG_xxx format which are mainly used for config settings required in makefile scripts and eventually also source files.
```
I changed to use MAX_MMAP_REGIONS.

So @jenswi-linaro @jforissier, you agree to use the following, add platform could define CFG_MMAP_REGIONS?
```
#ifndef CFG_MMAP_REGIONS
#define CFG_MMAP_REGIONS 13
#endif
#define MAX_MMAP_REGIONS CFG_MMAP_REGIONS
```

jforissier replied:
First thing: generally speaking, `#ifndef CFG_X / #define CFG_X / #endif` is not good. `CFG_x ?= some_value` in some `.mk` file should always be preferred.

Second thing: whether or not the variable should be a `CFG_` is easy to decide. If it makes sense to choose the value a compile time, use CFG_. Otherwise, hard-code the value in a .h and do NOT use #ifndef.

Third thing: I don't understand why you would need two variables (MAX_MMAP_REGIONS and CFG_MMAP_REGIONS).

It seems to me `CFG_MMAP_REGIONS ?= some_value` is all we need. It may be set globally in `core/config.mk` and a platform-specific value may be given in `core/arch/arm/plat-imx/conf.mk`. A compile-time assert may be in order to make sure the value is acceptable in the end.

MrVan replied:
Thanks @jforissier . Replacing MAX_MMAP_REGIONS with CFG_MMAP_REGIONS is simpler. How about the following patch?
```
diff --git a/core/arch/arm/arm.mk b/core/arch/arm/arm.mk
index bb65f77..6e52deb 100644
--- a/core/arch/arm/arm.mk
+++ b/core/arch/arm/arm.mk
@@ -4,6 +4,8 @@ CFG_LTC_OPTEE_THREAD ?= y
 CFG_CORE_TZSRAM_EMUL_SIZE ?= 458752
 CFG_LPAE_ADDR_SPACE_SIZE ?= (1ull << 32)
 
+CFG_MMAP_REGIONS ?= 13
+
 ifeq ($(CFG_ARM64_core),y)
 CFG_KERN_LINKER_FORMAT ?= elf64-littleaarch64
 CFG_KERN_LINKER_ARCH ?= aarch64
diff --git a/core/arch/arm/mm/core_mmu.c b/core/arch/arm/mm/core_mmu.c
index f2d728d..7c5fb3d 100644
--- a/core/arch/arm/mm/core_mmu.c
+++ b/core/arch/arm/mm/core_mmu.c
@@ -51,7 +51,6 @@
 
 #include "core_mmu_private.h"
 
-#define MAX_MMAP_REGIONS       13
 #define RES_VASPACE_SIZE       (CORE_MMU_PGDIR_SIZE * 10)
 #define SHM_VASPACE_SIZE       (1024 * 1024 * 32)
 
@@ -66,7 +65,7 @@ unsigned long default_nsec_shm_size;
 unsigned long default_nsec_shm_paddr;
 
 static struct tee_mmap_region
-       static_memory_map[MAX_MMAP_REGIONS + 1];
+       static_memory_map[CFG_MMAP_REGIONS + 1];
 
 /* Define the platform's memory layout. */
 struct memaccess_area {
@@ -911,6 +910,7 @@ void core_init_mmu_map(void)
                        panic("Invalid memory access config: sec/nsec");
        }
 
+       COMPILE_TIME_ASSERT(CFG_MMAP_REGIONS < 13);
        init_mem_map(static_memory_map, ARRAY_SIZE(static_memory_map));
 
        map = static_memory_map;
```

jforissier replied:
@MrVan looks good to me!

etienne-lms replied:
I think `MAX_MMAP_REGION` does not need to reach make script.
This value allocated static memory so it's a good thing to have it tunable. But the amount of memory is quite low. Maybe we would need a debug trace trace to get the number of entry used against max defined value.

In this scope, we should rather look at `MAX_XLAT_TABLES` which cover bigger areas.

Minor comment: why restrict to 13. If no valid value, will panic. Ok 13 to be a default value, but why the lowest supported ?

MrVan replied:
@etienne-lms I have no idea about the default value 13, 13 is not large enough in my case. The patch I listed above is the easy way to fix the issue I met.

Covering bigger areas, do you mean `XLAT_TABLE_ENTRIES`?

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -58,6 +58,15 @@ const struct dt_driver *__dt_driver_end(void)
+
+	prop = fdt_getprop(fdt, offs, propname, NULL);
+
+	return prop ? true : false;
```
There's no need for the ternary operator here. Since `dt_have_prop()` returns `bool` the compiler is required to normalize the value to return either `true` or `false`.

MrVan replied:
I am not sure, but let compiler to normalize a value to true or false maybe not good, different compilers may have different behaviors? gcc/llvm or else. I checked linux and freebsd code, they not use compiler to normalize the value. freebsd not use bool, see https://github.com/freebsd/freebsd/blob/master/sys/dev/ofw/openfirm.c#L371
Linux kernel 
```
static inline bool of_property_read_bool(const struct device_node *np,
                                           const char *propname)
  {
          struct property *prop = of_find_property(np, propname, NULL);
          return prop ? true : false;
  }
```

So, could we keep `prop ? true : false` here?

jforissier replied:
`? true : false` is clearly useless here because the `bool` type in `optee_os/core` is always defined as `_Bool`, which is C99 standard.
There are reaons why the links you provide have the normalization to true/false (or 0/1):
- The FreeBSD function returns `int`, which is not `_Bool`,
- The Linux function is part of [`libfdt`](https://github.com/dgibson/dtc/tree/master/libfdt), which makes no assumption on the actual type behind `bool`.

jenswi-linaro replied:
From the C99 standard:
*6.3.1.2 Boolean type
When any scalar value is converted to _Bool, the result is 0 if the value compares equal to 0; otherwise, the result is 1.*

MrVan replied:
Thanks @jforissier for the explaination. I'll fix it in next version.

MrVan commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -90,6 +90,11 @@ uint32_t sem_cpu_sync[CFG_TEE_CORE_NB_CORE];
 
+#ifdef CFG_DT
+void *dt_blob_addr;
+KEEP_PAGER(dt_blob_addr);
```
ok. Will remove it.

MrVan commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -378,6 +383,14 @@ static void init_runtime(unsigned long pageable_part __unused)
+	if (cpu_mmu_enabled())
+		return dt_blob_addr;
+
+	return (void *)virt_to_phys(dt_blob_addr);
```
ok. Then let the function directly `return dt_blob_addr` is ok here.

etienne-lms replied:
add `assert(cpu_mmu_enabled());`

MrVan replied:
ok. Will add it.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -760,6 +777,9 @@ static void init_primary_helper(unsigned long pageable_part,
 	if (init_teecore() != TEE_SUCCESS)
 		panic();
+#ifdef CFG_DT
+	dt_blob_addr = NULL;
```
Proposal to add comment `/* dt no more reached, reset pointer to invalid */`. It is not obvious why this is cleared.
Even add a `reset_dt_references();` to remove `#ifdef CFG_xxx`.

MrVan replied:
ok. I'll fix it.

etienne-lms commented on core/drivers/imx_wdog.c:
```diff
@@ -0,0 +1,160 @@
+#include <io.h>
+#include <keep.h>
+#include <kernel/generic_boot.h>
+#include <kernel/dt.h>
```
minor: 2 sorting error.

etienne-lms commented on core/drivers/imx_wdog.c:
```diff
@@ -0,0 +1,160 @@
+	paddr_t pbase;
+	vaddr_t vbase;
+	ssize_t sz;
+	int off, st;
```
minor: split in 2 lines.

MrVan commented:
Updated:
Remove KEEP_PAGER(dt_blob_addr)
Add static to dt_blob_addr
add "assert(cpu_mmu_enabled())" to get_dt_blob
Use CFG_MMAP_REGIONS to replace MAX_MMAP_REGIONS
Change "return prop ? true: fase" to "return prop"
Fix header sorting error
Split "int off, st" to two lines
Fix indention
Add a reset_dt_references to clear dt_blob_addr and remove ifdef CFG_XXX

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -90,6 +90,10 @@ uint32_t sem_cpu_sync[CFG_TEE_CORE_NB_CORE];
 #endif
 
+#ifdef CFG_DT
+void *dt_blob_addr;
```
This should be static

MrVan commented:
R-b Tags applied with changing"void *dt_blob_addr;" -> "static void *dt_blob_addr".
